### PR TITLE
 chore(engine): introduce blockchain tree action

### PR DESCRIPTION
**Pull request by rkrasiuk**
_It will get worse before it gets better._ - Sun Tzu

## Description

Builds on top of https://github.com/paradigmxyz/reth/pull/8027.
Introduces blockchain tree action which will gradually help detach blockchain tree calls from blockchain tree result processing.
Implements first blockchain tree action - make canonical call on FCU.

### CODE REVIEW

**Comment by rkrasiuk** [comment]
hive run https://github.com/paradigmxyz/reth/actions/runs/8912503678

File: crates/consensus/beacon/src/engine/mod.rs [hunk 1]
```diff
@@ -1719,34 +1774,7 @@ where
...
-                                    let _ = tx.send(Ok(on_updated));
-
-                                    if fcu_status.is_valid() {
-                                        let tip_number = this.blockchain.canonical_tip().number;
-                                        if this.sync.has_reached_max_block(tip_number) {
-                                            // Terminate the sync early if it's reached the
-                                            // maximum user configured block.
-                                            return Poll::Ready(Ok(()))
-                                        }
-                                    }
```
Comment (rkrasiuk) [inline crates/consensus/beacon/src/engine/mod.rs:1737]: fcu is never valid if the status is returned during pre-validation

**Review by onbjerg — APPROVED** [review]

File: crates/consensus/beacon/src/engine/mod.rs [hunk 2]
```diff
@@ -1711,6 +1756,16 @@ where
...
                 }
 
+                // Process any blockchain tree action result as set forth during engine message
+                // processing.
+                if let Some(action) = this.blockchain_tree_action.take() {
+                    match this.on_blockchain_tree_action(action)? {
+                        EngineEventOutcome::Processed => {}
+                        EngineEventOutcome::ReachedMaxBlock => return Poll::Ready(Ok(())),
+                    };
+                    continue
```
Comment (mattsse) [inline crates/consensus/beacon/src/engine/mod.rs:1766]:
```suggestion
```

**Review by mattsse — APPROVED** [review]
I get the idea,

this makes sense and with the way the poll loop works, FCUs are still executed right away

**Pull request merged**